model/note/note.go

```diff
@@ -222,40 +281,173 @@ func ensureNotesDir(inst *instance.Instance) (*vfs.DirDoc, error) {
+}
+
+// GetFile takes a file from the VFS as a note and returns its last version. It
+// is useful when some changes have not yet been persisted to the VFS.
+func GetFile(inst *instance.Instance, file *vfs.FileDoc) (*vfs.FileDoc, error) {
 	lock := inst.NotesLock()
 	if err := lock.Lock(); err != nil {
-		return err
+		return nil, err
 	}
 	defer lock.Unlock()
+	doc, err := get(inst, file)
+	if err != nil {
+		return nil, err
+	}
+	return doc.asFile(file), nil
+}
 
-	if len(file.Metadata) == 0 {
-		return ErrInvalidFile
+// get must be called with the notes lock already acquired. It will try to load
+// the last version if a note from the cache, and if it fails, it will replay
+// the new steps on the file from the VFS.
+func get(inst *instance.Instance, file *vfs.FileDoc) (*Document, error) {
+	if doc := getFromCache(inst, file.ID()); doc != nil {
+		return doc, nil
 	}
-	old, _ := file.Metadata["title"].(string)
-	if old == title {
+	// TODO replay steps
+	return fromMetadata(file)
+}
+
+func fromMetadata(file *vfs.FileDoc) (*Document, error) {
+	var version int64
+	switch v := file.Metadata["version"].(type) {
+	case float64:
+		version = int64(v)
+	case int64:
+		version = v
+	default:
+		return nil, ErrInvalidFile
+	}
+	title, _ := file.Metadata["title"].(string)
+	schema, ok := file.Metadata["schema"].(map[string]interface{})
+	if !ok {
+		return nil, ErrInvalidFile
+	}
+	content, ok := file.Metadata["content"].(map[string]interface{})
+	if !ok {
+		return nil, ErrInvalidFile
+	}
+	return &Document{
+		DocID:      file.ID(),
+		DirID:      file.DirID,
+		Title:      title,
+		Version:    version,
+		SchemaSpec: schema,
+		RawContent: content,
+	}, nil
+}
+
+func getFromCache(inst *instance.Instance, noteID string) *Document {
+	cache := config.GetConfig().CacheStorage
+	buf, ok := cache.Get(cacheKey(inst, noteID))
+	if !ok {
+		return nil
+	}
+	var doc Document
+	if err := json.Unmarshal(buf, &doc); err != nil {
 		return nil
 	}
+	return &doc
+}
+
+func clearCache(inst *instance.Instance, noteID string) {
+	cache := config.GetConfig().CacheStorage
+	cache.Clear(cacheKey(inst, noteID))
+}
+
+func cacheKey(inst *instance.Instance, noteID string) string {
+	return fmt.Sprintf("note:%s:%s", inst.Domain, noteID)
+}
+
+func saveToCache(inst *instance.Instance, doc *Document) error {
+	cache := config.GetConfig().CacheStorage
+	buf, err := json.Marshal(doc)
+	if err != nil {
+		return err
+	}
+	cache.Set(cacheKey(inst, doc.ID()), buf, 30*time.Minute)
+	return nil
+}
+
+// UpdateTitle changes the title of a note and renames the associated file.
+func UpdateTitle(inst *instance.Instance, file *vfs.FileDoc, title string) (*vfs.FileDoc, error) {
+	lock := inst.NotesLock()
+	if err := lock.Lock(); err != nil {
+		return nil, err
+	}
+	defer lock.Unlock()
 
+	doc, err := get(inst, file)
+	if err != nil {
+		return nil, err
+	}
+
+	if doc.Title == title {
+		return file, nil
+	}
+	doc.Title = title
+	if err := saveToCache(inst, doc); err != nil {
```

Comment:
Maybe we could invalidate the cache and keep going if we fail to save the new document to cache?
It seems a bit overkill to bail out if we can't save a perfectly fine document to the cache.

Reply:
No, for the title, if we can't save to the cache, the change will be lost. So, it's better to return an error.

Reply:
Hum yes. I probably wrote this before you added cache.